folding-schemes/src/folding/protogalaxy/mod.rs

```diff
@@ -50,20 +129,97 @@ impl<C: CurveGroup> AllocVar<CommittedInstance<C>, C::ScalarField> for Committed
                 phi: NonNativeAffineVar::new_variable(cs.clone(), || Ok(u.phi), mode)?,
                 betas: Vec::new_variable(cs.clone(), || Ok(u.betas.clone()), mode)?,
                 e: FpVar::new_variable(cs.clone(), || Ok(u.e), mode)?,
-                u: FpVar::new_variable(cs.clone(), || Ok(u.u), mode)?,
                 x: Vec::new_variable(cs.clone(), || Ok(u.x.clone()), mode)?,
             })
         })
     }
 }
 
+impl<C: CurveGroup> R1CSVar<C::ScalarField> for CommittedInstanceVar<C> {
+    type Value = CommittedInstance<C>;
+
+    fn cs(&self) -> ConstraintSystemRef<C::ScalarField> {
+        self.phi
+            .cs()
+            .or(self.betas.cs())
+            .or(self.e.cs())
+            .or(self.x.cs())
+    }
+
+    fn value(&self) -> Result<Self::Value, SynthesisError> {
+        Ok(CommittedInstance {
+            phi: self.phi.value()?,
+            betas: self
+                .betas
+                .iter()
+                .map(|v| v.value())
+                .collect::<Result<_, _>>()?,
+            e: self.e.value()?,
+            x: self.x.iter().map(|v| v.value()).collect::<Result<_, _>>()?,
+        })
+    }
+}
+
+impl<C: CurveGroup> CommittedInstanceVar<C>
+where
+    C::ScalarField: Absorb,
+    C::BaseField: PrimeField,
+{
+    /// hash implements the committed instance hash compatible with the native implementation from
+    /// CommittedInstance.hash.
+    /// Returns `H(i, z_0, z_i, U_i)`, where `i` can be `i` but also `i+1`, and `U` is the
+    /// `CommittedInstance`.
+    /// Additionally it returns the vector of the field elements from the self parameters, so they
+    /// can be reused in other gadgets avoiding recalculating (reconstraining) them.
+    #[allow(clippy::type_complexity)]
+    pub fn hash(
+        self,
+        sponge: &PoseidonSpongeVar<CF1<C>>,
+        pp_hash: FpVar<CF1<C>>,
+        i: FpVar<CF1<C>>,
+        z_0: Vec<FpVar<CF1<C>>>,
+        z_i: Vec<FpVar<CF1<C>>>,
+    ) -> Result<(FpVar<CF1<C>>, Vec<FpVar<CF1<C>>>), SynthesisError> {
+        let mut sponge = sponge.clone();
+        let U_vec = self.to_sponge_field_elements()?;
+        sponge.absorb(&pp_hash)?;
+        sponge.absorb(&i)?;
+        sponge.absorb(&z_0)?;
+        sponge.absorb(&z_i)?;
+        sponge.absorb(&U_vec)?;
+        Ok((sponge.squeeze_field_elements(1)?.pop().unwrap(), U_vec))
+    }
+}
+
 #[derive(Clone, Debug)]
 pub struct Witness<F: PrimeField> {
     w: Vec<F>,
     r_w: F,
 }
 
-#[derive(Debug, Error, PartialEq)]
+impl<F: PrimeField> Witness<F> {
+    pub fn new(w: Vec<F>) -> Self {
+        // note: at the current version, we don't use the blinding factors and we set them to 0
+        // always.
+        Self { w, r_w: F::zero() }
+    }
+
+    pub fn commit<CS: CommitmentScheme<C>, C: CurveGroup<ScalarField = F>>(
+        &self,
+        params: &CS::ProverParams,
+        x: Vec<F>,
```

Comment:
I guess a owned `x` is inevitable here because it is going to be a field in `CommittedInstance`. I don't mind using a slice as a parameter, but cloning will still happen inside the method.

On the other hand, the only place where `commit` is called is: https://github.com/privacy-scaling-explorations/sonobe/blob/eca484912af75c180608dfee094047a96b38fd79/folding-schemes/src/folding/protogalaxy/mod.rs#L812

As we can see, the ownership of `x_i1` is transferred to the method, while no memory allocation is needed.

Reply:
That makes sense.
Missed it!